File: src/stores/channel.ts
import { defineStore } from "pinia";
import {
  createChannel,
  deleteChannel,
  getChannelRuntimeStatus,
  listChannels,
  startChannelRuntime,
  stopChannelRuntime,
  testChannel,
  updateChannel,
  type ChannelDraft,
  type ChannelRecord,
  type ChannelRuntimeStatusRecord,
  type ChannelTestReport,
  type ChannelTestResult
} from "@/api/tauri";

export interface ChannelItem {
  id: string;
  name: string;
  kind: string;
  configJson: string;
  enabled: boolean;
  lastCheckedAt: string | null;
  lastHealthStatus: string | null;
  lastHealthMessage: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface ChannelTestItem {
  ok: boolean;
  kind: string;
  message: string;
  checkedAt: string;
}

export interface ChannelRuntimeStatusItem {
  running: boolean;
  state: string;
  message: string;
  updatedAt: string;
}

function normalizeTime(value: string | null | undefined) {
  if (!value) {
    return 0;
  }

  if (/^\d+$/.test(value)) {
    return Number(value);
  }

  return Number(new Date(value)) || 0;
}

function mapChannel(record: ChannelRecord): ChannelItem {
  return {
    id: record.id,
    name: record.name,
    kind: record.kind,
    configJson: record.config_json,
    enabled: record.enabled,
    lastCheckedAt: record.last_checked_at,
    lastHealthStatus: record.last_health_status,
    lastHealthMessage: record.last_health_message,
    createdAt: record.created_at,
    updatedAt: record.updated_at
  };
}

function mapTestReport(report: ChannelTestReport): ChannelTestItem {
  return {
    ok: report.ok,
    kind: report.kind,
    message: report.message,
    checkedAt: report.checked_at
  };
}

function mapRuntimeStatus(status: ChannelRuntimeStatusRecord): ChannelRuntimeStatusItem {
  return {
    running: status.running,
    state: status.state,
    message: status.message,
    updatedAt: status.updated_at
  };
}

function sortChannels(channels: ChannelItem[]) {
  return [...channels].sort((left, right) => {
    if (left.enabled !== right.enabled) {
      return left.enabled ? -1 : 1;
    }

    return normalizeTime(right.updatedAt) - normalizeTime(left.updatedAt);
  });
}

function upsertChannel(channels: ChannelItem[], channel: ChannelItem) {
  const exists = channels.some((item) => item.id === channel.id);
  return sortChannels(exists ? channels.map((item) => (item.id === channel.id ? channel : item)) : [channel, ...channels]);
}

export const defaultRuntimeStatus = (): ChannelRuntimeStatusItem => ({
  running: false,
  state: "idle",
  message: "Channels supervisor is idle.",
  updatedAt: ""
});

export const useChannelStore = defineStore("channels", {
  state: () => ({
    channels: [] as ChannelItem[],
    activeChannelId: "" as string,
    loaded: false,
    isLoading: false,
    isSaving: false,
    isTesting: false,
    isStartingRuntime: false,
    isStoppingRuntime: false,
    error: "" as string,
    runtimeStatus: defaultRuntimeStatus() as ChannelRuntimeStatusItem,
    lastTestReport: null as ChannelTestItem | null
  }),
  getters: {
    activeChannel(state) {
      return state.channels.find((channel) => channel.id === state.activeChannelId) ?? null;
    },
    enabledCount(state) {
      return state.channels.filter((channel) => channel.enabled).length;
    },
    healthyCount(state) {
      return state.channels.filter((channel) => channel.lastHealthStatus === "healthy").length;
    }
  },
  actions: {
    setActiveChannel(channelId: string) {
      this.activeChannelId = channelId;
    },
    clearActiveChannel() {
      this.activeChannelId = "";
      this.lastTestReport = null;
    },
    applyChannel(record: ChannelRecord) {
      const channel = mapChannel(record);
      this.channels = upsertChannel(this.channels, channel);
      if (!this.activeChannelId) {
        this.activeChannelId = channel.id;
      }
      return channel;
    },
    applyRuntimeStatus(status: ChannelRuntimeStatusRecord | ChannelRuntimeStatusItem) {
      this.runtimeStatus = "updated_at" in status ? mapRuntimeStatus(status) : status;
      return this.runtimeStatus;
    },
    async bootstrap() {
      if (this.loaded || this.isLoading) {
        return this.channels;
      }

      this.isLoading = true;
      this.error = "";

      try {
        const [channels, runtimeStatus] = await Promise.all([listChannels(), getChannelRuntimeStatus()]);
        this.channels = sortChannels(channels.map(mapChannel));
        this.runtimeStatus = mapRuntimeStatus(runtimeStatus);
        this.loaded = true;
        if (!this.activeChannelId && this.channels.length > 0) {
          this.activeChannelId = this.channels[0].id;
        }
        return this.channels;
      } catch (error) {
        this.error = error instanceof Error ? error.message : String(error);
        throw error;
      } finally {
        this.isLoading = false;
      }
    },
    async createChannel(channel: ChannelDraft) {
      this.isSaving = true;
      this.error = "";

      try {
        const created = this.applyChannel(await createChannel(channel));
        this.activeChannelId = created.id;
        this.loaded = true;
        return created;
      } catch (error) {
        this.error = error instanceof Error ? error.message : String(error);
        throw error;
      } finally {
        this.isSaving = false;
      }
    },
    async updateChannel(channelId: string, channel: ChannelDraft) {
      this.isSaving = true;
      this.error = "";

      try {
        const updated = this.applyChannel(await updateChannel(channelId, channel));
        this.activeChannelId = updated.id;
        return updated;
      } catch (error) {
        this.error = error instanceof Error ? error.message : String(error);
        throw error;
      } finally {
        this.isSaving = false;
      }
    },
    async deleteChannel(channelId: string) {
      this.isSaving = true;
      this.error = "";

      try {
        await deleteChannel(channelId);
        this.channels = this.channels.filter((channel) => channel.id !== channelId);
        if (this.activeChannelId === channelId) {
          this.activeChannelId = this.channels[0]?.id ?? "";
        }
      } catch (error) {
        this.error = error instanceof Error ? error.message : String(error);
        throw error;
      } finally {
        this.isSaving = false;
      }
    },
    async testChannel(channelId: string) {
      this.isTesting = true;
      this.error = "";
      this.lastTestReport = null;

      try {
        const result = await testChannel(channelId);
        this.applyChannel(result.channel);
        this.lastTestReport = mapTestReport(result.report);
        return result as ChannelTestResult;
      } catch (error) {
        this.error = error instanceof Error ? error.message : String(error);
        throw error;
      } finally {
        this.isTesting = false;
      }
    },
    async startRuntime() {
      this.isStartingRuntime = true;
      this.error = "";

      try {
        this.runtimeStatus = mapRuntimeStatus(await startChannelRuntime());
        return this.runtimeStatus;
      } catch (error) {
        this.error = error instanceof Error ? error.message : String(error);
        throw error;
      } finally {
        this.isStartingRuntime = false;
      }
    },
    async stopRuntime() {
      this.isStoppingRuntime = true;
      this.error = "";

      try {
        this.runtimeStatus = mapRuntimeStatus(await stopChannelRuntime());
        return this.runtimeStatus;
      } catch (error) {
        this.error = error instanceof Error ? error.message : String(error);
        throw error;
      } finally {
        this.isStoppingRuntime = false;
      }
    }
  }
});

File: src-tauri/src/services/channel.rs
use std::{sync::Arc, time::Duration};

use chrono::Utc;
use tauri::{AppHandle, Emitter};
use zeroclaw::{
    channels::{
        self, Channel as RuntimeChannel, DiscordChannel, SlackChannel, TelegramChannel,
        WebhookChannel,
    },
    config::{ChannelsConfig, DiscordConfig, SlackConfig, TelegramConfig, WebhookConfig},
};

use crate::{
    db,
    models::channel::{
        ChannelDraft, ChannelRecord, ChannelRuntimeStatusRecord, ChannelTestReport,
        ChannelTestResult,
    },
    state::AppState,
};

const CHANNELS_RUNTIME_EVENT: &str = "channels:runtime-status";

pub fn list_channels(state: &AppState) -> Result<Vec<ChannelRecord>, String> {
    db::list_channels(&state.db_path())
}

pub fn create_channel(state: &AppState, channel: ChannelDraft) -> Result<ChannelRecord, String> {
    let normalized = normalize_channel_draft(channel)?;
    db::upsert_channel(
        &state.db_path(),
        None,
        &normalized.name,
        &normalized.kind,
        &normalized.config_json,
        normalized.enabled,
    )
}

pub fn update_channel(
    state: &AppState,
    channel_id: &str,
    channel: ChannelDraft,
) -> Result<ChannelRecord, String> {
    let normalized = normalize_channel_draft(channel)?;
    db::upsert_channel(
        &state.db_path(),
        Some(channel_id),
        &normalized.name,
        &normalized.kind,
        &normalized.config_json,
        normalized.enabled,
    )
}

pub fn delete_channel(state: &AppState, channel_id: &str) -> Result<(), String> {
    db::delete_channel(&state.db_path(), channel_id)
}

pub async fn test_channel(state: &AppState, channel_id: &str) -> Result<ChannelTestResult, String> {
    let record = db::get_channel(&state.db_path(), channel_id)?
        .ok_or_else(|| "Channel not found.".to_string())?;
    let report = run_health_check(&record.kind, &record.config_json).await?;

    let updated = db::update_channel_health(
        &state.db_path(),
        &record.id,
        &report.checked_at,
        if report.ok { "healthy" } else { "error" },
        &report.message,
    )?;

    Ok(ChannelTestResult {
        channel: updated,
        report,
    })
}

pub fn get_runtime_status(state: &AppState) -> ChannelRuntimeStatusRecord {
    state.get_channel_runtime_status()
}

pub async fn start_runtime(
    app: AppHandle,
    state: AppState,
) -> Result<ChannelRuntimeStatusRecord, String> {
    if state.has_channel_runtime_task() {
        return Ok(state.get_channel_runtime_status());
    }

    let config = build_runtime_config_with_channels(&state)?;
    if !has_enabled_channels(&config.channels_config) {
        return Err(
            "Enable at least one configured channel before starting the runtime.".to_string(),
        );
    }

    let starting = make_runtime_status(true, "starting", "Launching the channels supervisor.");
    emit_runtime_status(&app, &state, starting.clone())?;

    let app_handle = app.clone();
    let state_clone = state.clone();
    let task = tauri::async_runtime::spawn(async move {
        let running = make_runtime_status(true, "running", "Channels supervisor is running.");
        let _ = emit_runtime_status(&app_handle, &state_clone, running);

        let result = channels::start_channels(config).await;
        state_clone.clear_channel_runtime_task();

        let final_status = match result {
            Ok(_) => make_runtime_status(false, "stopped", "Channels supervisor exited."),
            Err(error) => make_runtime_status(
                false,
                "error",
                &format!("Channels supervisor failed: {error}"),
            ),
        };

        let _ = emit_runtime_status(&app_handle, &state_clone, final_status);
    });

    state.store_channel_runtime_task(task);
    Ok(state.get_channel_runtime_status())
}

pub fn stop_runtime(
    app: AppHandle,
    state: &AppState,
) -> Result<ChannelRuntimeStatusRecord, String> {
    if let Some(task) = state.take_channel_runtime_task() {
        task.abort();
    }

    let status = make_runtime_status(false, "stopped", "Channels supervisor stopped by user.");
    emit_runtime_status(&app, state, status.clone())?;
    Ok(status)
}

fn normalize_channel_draft(channel: ChannelDraft) -> Result<ChannelDraft, String> {
    let kind = channel.kind.trim().to_lowercase();
    if channel.name.trim().is_empty() {
        return Err("Channel name cannot be empty.".to_string());
    }

    let config_json = normalize_config_json(&kind, &channel.config_json)?;
    validate_channel_kind(&kind)?;

    Ok(ChannelDraft {
        name: channel.name.trim().to_string(),
        kind,
        config_json,
        enabled: channel.enabled,
    })
}

async fn run_health_check(kind: &str, config_json: &str) -> Result<ChannelTestReport, String> {
    let checked_at = Utc::now().timestamp_millis().to_string();
    let channel = build_channel(kind, config_json)?;

    let result = tokio::time::timeout(Duration::from_secs(10), channel.health_check()).await;
    let (ok, message) = match result {
        Ok(true) => (true, "Health check passed.".to_string()),
        Ok(false) => (
            false,
            "Health check failed. Verify credentials, permissions, and network access.".to_string(),
        ),
        Err(_) => (
            false,
            "Health check timed out after 10 seconds.".to_string(),
        ),
    };

    Ok(ChannelTestReport {
        ok,
        kind: kind.to_string(),
        message,
        checked_at,
    })
}

fn build_runtime_config_with_channels(
    state: &AppState,
) -> Result<zeroclaw::config::Config, String> {
    let settings = super::runtime::load_runtime_settings(&state.settings_path())?;
    let mut config = super::runtime::build_resolved_runtime_config(&state.db_path(), settings)?;
    config.channels_config = build_channels_config(&state.db_path())?;
    Ok(config)
}

fn build_channels_config(db_path: &std::path::Path) -> Result<ChannelsConfig, String> {
    let mut config = ChannelsConfig::default();
    config.cli = false;

    for record in db::list_channels(db_path)? {
        if !record.enabled {
            continue;
        }

        match record.kind.as_str() {
            "telegram" => {
                config.telegram = Some(parse_json_config::<TelegramConfig>(&record.config_json)?);
            }
            "discord" => {
                config.discord = Some(parse_json_config::<DiscordConfig>(&record.config_json)?);
            }
            "slack" => {
                config.slack = Some(parse_json_config::<SlackConfig>(&record.config_json)?);
            }
            "webhook" => {
                config.webhook = Some(parse_json_config::<WebhookConfig>(&record.config_json)?);
            }
            _ => {}
        }
    }

    Ok(config)
}

fn has_enabled_channels(config: &ChannelsConfig) -> bool {
    config.telegram.is_some()
        || config.discord.is_some()
        || config.slack.is_some()
        || config.webhook.is_some()
}

fn build_channel(kind: &str, config_json: &str) -> Result<Arc<dyn RuntimeChannel>, String> {
    match kind {
        "telegram" => {
            let config = parse_json_config::<TelegramConfig>(config_json)?;
            Ok(Arc::new(TelegramChannel::new(
                config.bot_token,
                config.allowed_users,
                config.mention_only,
            )))
        }
        "discord" => {
            let config = parse_json_config::<DiscordConfig>(config_json)?;
            Ok(Arc::new(DiscordChannel::new(
                config.bot_token,
                config.guild_id,
                config.allowed_users,
                config.listen_to_bots,
                config.mention_only,
            )))
        }
        "slack" => {
            let config = parse_json_config::<SlackConfig>(config_json)?;
            Ok(Arc::new(
                SlackChannel::new(
                    config.bot_token,
                    config.app_token,
                    config.channel_id,
                    Vec::new(),
                    config.allowed_users,
                )
                .with_group_reply_policy(config.mention_only, Vec::new()),
            ))
        }
        "webhook" => {
            let config = parse_json_config::<WebhookConfig>(config_json)?;
            Ok(Arc::new(WebhookChannel::new(
                config.port,
                config.listen_path,
                config.send_url,
                config.send_method,
                config.auth_header,
                config.secret,
            )))
        }
        _ => Err(
            "Unsupported channel kind. Supported kinds: telegram, discord, slack, webhook."
                .to_string(),
        ),
    }
}

fn normalize_config_json(kind: &str, config_json: &str) -> Result<String, String> {
    match kind {
        "telegram" => normalize_json_config::<TelegramConfig>(config_json),
        "discord" => normalize_json_config::<DiscordConfig>(config_json),
        "slack" => normalize_json_config::<SlackConfig>(config_json),
        "webhook" => normalize_json_config::<WebhookConfig>(config_json),
        _ => Err(
            "Unsupported channel kind. Supported kinds: telegram, discord, slack, webhook."
                .to_string(),
        ),
    }
}

fn validate_channel_kind(kind: &str) -> Result<(), String> {
    match kind {
        "telegram" | "discord" | "slack" | "webhook" => Ok(()),
        _ => Err(
            "Unsupported channel kind. Supported kinds: telegram, discord, slack, webhook."
                .to_string(),
        ),
    }
}

fn parse_json_config<T>(config_json: &str) -> Result<T, String>
where
    T: for<'de> serde::Deserialize<'de>,
{
    serde_json::from_str(config_json)
        .map_err(|error| format!("Failed to parse channel config JSON: {error}"))
}

fn normalize_json_config<T>(config_json: &str) -> Result<String, String>
where
    T: for<'de> serde::Deserialize<'de> + serde::Serialize,
{
    let parsed = parse_json_config::<T>(config_json)?;
    serde_json::to_string_pretty(&parsed).map_err(|error| error.to_string())
}

fn make_runtime_status(running: bool, state: &str, message: &str) -> ChannelRuntimeStatusRecord {
    ChannelRuntimeStatusRecord {
        running,
        state: state.to_string(),
        message: message.to_string(),
        updated_at: Utc::now().timestamp_millis().to_string(),
    }
}

fn emit_runtime_status(
    app: &AppHandle,
    state: &AppState,
    status: ChannelRuntimeStatusRecord,
) -> Result<(), String> {
    state.set_channel_runtime_status(status.clone());
    app.emit(CHANNELS_RUNTIME_EVENT, &status)
        .map_err(|error| error.to_string())
}
